java/src/main/java/com/google/oak/transport/GrpcStreamingTransport.java

```diff
@@ -0,0 +1,150 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import oak.session.noninteractive.v1.AttestationBundle;
+import oak.session.noninteractive.v1.GetPublicKeyRequest;
+import oak.session.noninteractive.v1.GetPublicKeyResponse;
+import oak.session.noninteractive.v1.InvokeRequest;
+import oak.session.noninteractive.v1.InvokeResponse;
+import oak.session.noninteractive.v1.RequestWrapper;
+import oak.session.noninteractive.v1.ResponseWrapper;
+
+public class GrpcStreamingTransport implements EvidenceProvider, Transport {
+  private static final Logger logger = Logger.getLogger(GrpcStreamingTransport.class.getName());
+
+  private static final int MESSAGE_QUEUE_TIMEOUT_SECONDS = 10;
+
+  private class ResponseStreamObserver implements StreamObserver<ResponseWrapper> {
+    @Override
+    public void onNext(ResponseWrapper response) {
+      try {
+        messageQueue.put(response);
+      } catch (Exception e) {
+        if (e instanceof InterruptedException) {
+          Thread.currentThread().interrupt();
+        }
+        logger.log(Level.WARNING, "Couldn't send server response to the message queue: {0}", e);
+      }
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      Status status = Status.fromThrowable(t);
+      logger.log(Level.WARNING, "Couldn't receive response: {0}", status);
+    }
+
+    @Override
+    public void onCompleted() {
+      logger.log(Level.FINE, "response message queue completed");
+    }
+  }
+
+  private BlockingQueue<ResponseWrapper> messageQueue;
+  private StreamObserver<RequestWrapper> requestObserver;
+
+  /**
+   * Creates an instance of {@code GrpcStreamingTransport}.
+   *
+   * @param stream a method reference to a gRPC client streaming method with the appropriate request
+   * and response types.
+   */
+  public GrpcStreamingTransport(
+      Function<StreamObserver<ResponseWrapper>, StreamObserver<RequestWrapper>> stream) {
+    // We expect to receive a single response message for each request.
+    this.messageQueue = new ArrayBlockingQueue<>(1);
+    StreamObserver<ResponseWrapper> responseObserver = new ResponseStreamObserver();
+    this.requestObserver = stream.apply(responseObserver);
+  }
+
+  /**
+   * Returns evidence about the trustworthiness of a remote server.
+   *
+   * @return {@code AttestationBundle} wrapped in a {@code Result}
+   */
+  public Result<AttestationBundle, String> getEvidence() {
+    RequestWrapper requestWrapper = RequestWrapper.newBuilder()
+                                        .setGetPublicKeyRequest(GetPublicKeyRequest.newBuilder())
+                                        .build();
+    logger.log(Level.INFO, "sending get public key request: {0}", requestWrapper.toString());
+    this.requestObserver.onNext(requestWrapper);
+
+    ResponseWrapper responseWrapper;
+    try {
+      // TODO(#3644): Add retry for client messages.
+      responseWrapper = this.messageQueue.poll(MESSAGE_QUEUE_TIMEOUT_SECONDS, SECONDS);
```

Comment:
This may return null. You should handle that case too, in the code that follows. I suggesting addng a check right after the try-catch block to check if `responseWrapper` is null, and return an error result if so.